<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 3.0//EN">
<!--Converted with LaTeX2HTML 96.1 (Feb 5, 1996) by Nikos Drakos (nikos@cbl.leeds.ac.uk), CBLU, University of Leeds -->
<HTML>
<HEAD>
<TITLE>2.2 Built-in Exceptions</TITLE>
<META NAME="description" CONTENT="2.2 Built-in Exceptions">
<META NAME="keywords" CONTENT="lib">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">
<LINK REL=STYLESHEET HREF="lib.css">
</HEAD>
<BODY LANG="EN">
 <A NAME="tex2html2541" HREF="node26.html"><IMG WIDTH=37 HEIGHT=24 ALIGN=BOTTOM ALT="next" SRC="../icons/next_motif.gif" BORDER=0></A> <A NAME="tex2html2539" HREF="node3.html"><IMG WIDTH=26 HEIGHT=24 ALIGN=BOTTOM ALT="up" SRC="../icons/up_motif.gif" BORDER=0></A> <A NAME="tex2html2533" HREF="node24.html"><IMG WIDTH=63 HEIGHT=24 ALIGN=BOTTOM ALT="previous" SRC="../icons/previous_motif.gif" BORDER=0></A> <A NAME="tex2html2543" HREF="node1.html"><IMG WIDTH=65 HEIGHT=24 ALIGN=BOTTOM ALT="contents" SRC="../icons/contents_motif.gif" BORDER=0></A> <A NAME="tex2html2544" HREF="node221.html"><IMG WIDTH=43 HEIGHT=24 ALIGN=BOTTOM ALT="index" SRC="../icons/index_motif.gif" BORDER=0></A> <BR>
<B> Next:</B> <A NAME="tex2html2542" HREF="node26.html">2.3 Built-in Functions</A>
<B>Up:</B> <A NAME="tex2html2540" HREF="node3.html">2 Built-in TypesExceptions </A>
<B> Previous:</B> <A NAME="tex2html2534" HREF="node24.html">2.1.8 Special Attributes</A>
<BR><HR><P>
<H1><A NAME="SECTION00320000000000000000">2.2 Built-in Exceptions</A></H1>
<P>
Exceptions are string objects.  Two distinct string objects with the
same value are different exceptions.  This is done to force programmers
to use exception names rather than their string value when specifying
exception handlers.  The string value of all built-in exceptions is
their name, but this is not a requirement for user-defined exceptions
or exceptions defined by library modules.
<P>
The following exceptions can be generated by the interpreter or
built-in functions.  Except where mentioned, they have an `associated
value' indicating the detailed cause of the error.  This may be a
string or a tuple containing several items of information (e.g., an
error code and a string explaining the code).
<P>
User code can raise built-in exceptions.  This can be used to test an
exception handler or to report an error condition `just like' the
situation in which the interpreter raises the same exception; but
beware that there is nothing to prevent user code from raising an
inappropriate error.
<P>
<DL><DT><STRONG><A NAME="">AttributeError</A></STRONG>
<DD>
Raised when an attribute reference or assignment fails.  (When an
  object does not support attribute references or attribute assignments
  at all, <CODE>TypeError</CODE> is raised.)
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">EOFError</A></STRONG>
<DD>
Raised when one of the built-in functions (<CODE>input()</CODE> or
  <CODE>raw_input()</CODE>) hits an end-of-file condition (EOF) without
  reading any data.
(N.B.: the <CODE>read()</CODE> and <CODE>readline()</CODE> methods of file
  objects return an empty string when they hit EOF.)  No associated value.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">IOError</A></STRONG>
<DD>
Raised when an I/O operation (such as a <CODE>print</CODE> statement, the
  built-in <CODE>open()</CODE> function or a method of a file object) fails
  for an I/O-related reason, e.g., `file not found', `disk full'.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">ImportError</A></STRONG>
<DD>
Raised when an <CODE>import</CODE> statement fails to find the module
  definition or when a <CODE>from ... import</CODE> fails to find a
  name that is to be imported.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">IndexError</A></STRONG>
<DD>
Raised when a sequence subscript is out of range.  (Slice indices are
  silently truncated to fall in the allowed range; if an index is not a
  plain integer, <CODE>TypeError</CODE> is raised.)
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">KeyError</A></STRONG>
<DD>
Raised when a mapping (dictionary) key is not found in the set of
  existing keys.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">KeyboardInterrupt</A></STRONG>
<DD>
  Raised when the user hits the interrupt key (normally
  <KBD>Control-C</KBD> or
<TT>DEL</TT>).  During execution, a check for interrupts is made regularly.
Interrupts typed when a built-in function <CODE>input()</CODE> or
  <CODE>raw_input()</CODE>) is waiting for input also raise this exception.  No
  associated value.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">MemoryError</A></STRONG>
<DD>
  Raised when an operation runs out of memory but the situation may
  still be rescued (by deleting some objects).  The associated value is
  a string indicating what kind of (internal) operation ran out of memory.
  Note that because of the underlying memory management architecture
  (C's <CODE>malloc()</CODE> function), the interpreter may not always be able
  to completely recover from this situation; it nevertheless raises an
  exception so that a stack traceback can be printed, in case a run-away
  program was the cause.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">NameError</A></STRONG>
<DD>
  Raised when a local or global name is not found.  This applies only
  to unqualified names.  The associated value is the name that could
  not be found.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">OverflowError</A></STRONG>
<DD>
Raised when the result of an arithmetic operation is too large to be
  represented.  This cannot occur for long integers (which would rather
  raise <CODE>MemoryError</CODE> than give up).  Because of the lack of
  standardization of floating point exception handling in C, most
  floating point operations also aren't checked.  For plain integers,
  all operations that can overflow are checked except left shift, where
  typical applications prefer to drop bits than raise an exception.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">RuntimeError</A></STRONG>
<DD>
  Raised when an error is detected that doesn't fall in any of the
  other categories.  The associated value is a string indicating what
  precisely went wrong.  (This exception is a relic from a previous
  version of the interpreter; it is not used any more except by some
  extension modules that haven't been converted to define their own
  exceptions yet.)
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">SyntaxError</A></STRONG>
<DD>
Raised when the parser encounters a syntax error.  This may occur in
  an <CODE>import</CODE> statement, in an <CODE>exec</CODE> statement, in a call
  to the built-in function <CODE>eval()</CODE> or <CODE>input()</CODE>, or
  when reading the initial script or standard input (also
  interactively).
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">SystemError</A></STRONG>
<DD>
  Raised when the interpreter finds an internal error, but the
  situation does not look so serious to cause it to abandon all hope.
  The associated value is a string indicating what went wrong (in
  low-level terms).
<P>
  You should report this to the author or maintainer of your Python
  interpreter.  Be sure to report the version string of the Python
  interpreter (<CODE>sys.version</CODE>; it is also printed at the start of an
  interactive Python session), the exact error message (the exception's
  associated value) and if possible the source of the program that
  triggered the error.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">SystemExit</A></STRONG>
<DD>
This exception is raised by the <CODE>sys.exit()</CODE> function.  When it
  is not handled, the Python interpreter exits; no stack traceback is
  printed.  If the associated value is a plain integer, it specifies the
  system exit status (passed to C's <CODE>exit()</CODE> function); if it is
  <CODE>None</CODE>, the exit status is zero; if it has another type (such as
  a string), the object's value is printed and the exit status is one.
<P>
  A call to <CODE>sys.exit</CODE> is translated into an exception so that
  clean-up handlers (<CODE>finally</CODE> clauses of <CODE>try</CODE> statements)
  can be executed, and so that a debugger can execute a script without
  running the risk of losing control.  The <CODE>posix._exit()</CODE> function
  can be used if it is absolutely positively necessary to exit
  immediately (e.g., after a <CODE>fork()</CODE> in the child process).
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">TypeError</A></STRONG>
<DD>
  Raised when a built-in operation or function is applied to an object
  of inappropriate type.  The associated value is a string giving
  details about the type mismatch.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">ValueError</A></STRONG>
<DD>
  Raised when a built-in operation or function receives an argument
  that has the right type but an inappropriate value, and the
  situation is not described by a more precise exception such as
  <CODE>IndexError</CODE>.
<P>
</DL>
<P>
<DL><DT><STRONG><A NAME="">ZeroDivisionError</A></STRONG>
<DD>
  Raised when the second argument of a division or modulo operation is
  zero.  The associated value is a string indicating the type of the
  operands and the operation.
<P>
</DL><HR><A NAME="tex2html2541" HREF="node26.html"><IMG WIDTH=37 HEIGHT=24 ALIGN=BOTTOM ALT="next" SRC="../icons/next_motif.gif" BORDER=0></A> <A NAME="tex2html2539" HREF="node3.html"><IMG WIDTH=26 HEIGHT=24 ALIGN=BOTTOM ALT="up" SRC="../icons/up_motif.gif" BORDER=0></A> <A NAME="tex2html2533" HREF="node24.html"><IMG WIDTH=63 HEIGHT=24 ALIGN=BOTTOM ALT="previous" SRC="../icons/previous_motif.gif" BORDER=0></A> <A NAME="tex2html2543" HREF="node1.html"><IMG WIDTH=65 HEIGHT=24 ALIGN=BOTTOM ALT="contents" SRC="../icons/contents_motif.gif" BORDER=0></A> <A NAME="tex2html2544" HREF="node221.html"><IMG WIDTH=43 HEIGHT=24 ALIGN=BOTTOM ALT="index" SRC="../icons/index_motif.gif" BORDER=0></A> <BR>
<B> Next:</B> <A NAME="tex2html2542" HREF="node26.html">2.3 Built-in Functions</A>
<B>Up:</B> <A NAME="tex2html2540" HREF="node3.html">2 Built-in TypesExceptions </A>
<B> Previous:</B> <A NAME="tex2html2534" HREF="node24.html">2.1.8 Special Attributes</A>
<P><ADDRESS>
guido@cnri.reston.va.us
</ADDRESS>
</BODY>
</HTML>
